### DISCUSSION TITLE
Dicas de livros

### DISCUSSION THREAD

**Discussion by manzoliric** [backend-br/forum#127] [Geral]
Estava pensando em criar uma Issue fixa com dicas de Livros e Videos voltados para cada assunto, exemplo.

PHP
Livro 1 - Assunto 1.

Python
Livro 1 - Assunto 1.

O que acham?

Legal ter uma topico fixo assim, porque existem muitos livros e com indicações de quem já leu fica mais facil escolher e tambem não comprar um livro já ultrapassado.

**Comment by reginaldojunior** [comment]
Clean Code - Código Limpo.

https://www.saraiva.com.br/cdigo-limpo-habilidades-prticas-do-agile-software-edio-revisada-2666082.html?pac_id=123134&gclid=CjwKCAjwrO_MBRBxEiwAYJnDLPBN-umurrkh79n92HIlOgMJFxk2nXvgC34x5eup3-Y349esomnkkBoCI1cQAvD_BwE

Acho que este é um livro que todos desenvolvedores devem ler pelo menos uma vez.

**Comment by diegocoxta** [comment]
Meus livros favoritos:
- Clean Code do Robert Cecil Martin (Como já citado pelo @reginaldojunior )
- Padrões de Projeto - Soluções Reutilizaveis de Software Orientado a Objetos dos GoF
- O Programador Pragmático: De Aprendiz a Mestre de Andy Hunt e Dave Thomas

**Reply by igorpereirag** [reply]
Muito top esses

**Comment by reginaldojunior** [comment]
Sobre Livros,

O melhor site de livro da internet (http://livrosemapego.com.br/livros) e é opensource <3

https://github.com/mtsalcova/livro-sem-apego

**Comment by khaueviana** [comment]
TDD - Kent Beck
TDD - Mauricio Aniche
Design Patterns - Use a cabeça
DDD - Erick Evans
DDD - Vernon
Padrões de Projetos do GoF (Erich Gamma e a rapaziada)
Padrões de Implementação - Kent Beck
Refatoração - Martin Fowler
Padrões de aplicações empresariais - Martin Fowler 
Clean Code - Uncle Bob

Tenho quase todos, mas não li todos haha

**Reply by igorpereirag** [reply]
Esses são incríveis

**Comment by woliveiras** [comment]
Fiz algumas listas desse tipo:

- [Livros que todo desenvolvedor de software deveria ler](https://woliveiras.com.br/posts/Livros-que-todo-desenvolvedor-de-software-deveria-ler/)
- [Livros sobre JavaScript, do iniciante ao avançado e ES6](https://woliveiras.com.br/posts/Livros-sobre-JavaScript-do-iniciante-ao-avancado-e-ES6/)
- [Vim para Noobs](https://woliveiras.com.br/vimparanoobs/) :heart_eyes_cat: 

Tá! O Vim para Noobs é zoeira de minha parte, não é requisito para nenhum dev.

**Comment by diegocoxta** [comment]
@woliveiras Muito boa a lista de livros JS.

**Comment by khaueviana** [comment]
Excelente!!!

**Comment by marianaarosa** [comment]
Olá... bom eu indico esse
**PYTHON**
Introdução à programação com Python
Do: Nilo Ney Coutinho Menezes. Ed. Novatec.

**Comment by opchaves** [comment]
Eu indico este repositório que contém centenas de indicações.

https://github.com/EbookFoundation/free-programming-books

**Comment by onhernandes** [comment]
Type Theory
Tem bons livros aqui:
https://github.com/jozefg/learn-tt

Esse aqui é ótimo pra algoritmos: https://www.amazon.com/Introduction-Algorithms-3rd-MIT-Press/dp/0262033844/ref=sr_1_1?ie=UTF8&qid=1508511454&sr=8-1&keywords=algorithms

E se quiser assistir algumas video-aulas sobre: http://purelytheoretical.com/sywtltt.html

**Comment by wilsonjuniordev** [comment]
Não são livros técnicos, mas bem interessante para quem está começando na área. 

- Guia do mestre programador Pensando como pirata, evoluindo como jedi
- O Programador Apaixonado Construindo uma carreira notável em desenvolvimento de software

**Comment by igorpereirag** [comment]
"Clean Code" por Robert C. Martin - Um guia prático para escrever código limpo e de alta qualidade."Java: The Complete Reference" por Herbert Schildt - Uma referência abrangente para programação em Java, desde conceitos básicos até tópicos avançados."Node.js Design Patterns" por Mario Casciaro e Luciano Mammino - Explora padrões de design e melhores práticas para desenvolvimento back-end com Node.js.Esses livros fornecem uma base sólida para quem deseja aprender e aprimorar suas habilidades no desenvolvimento back-end

**Comment by hebcodee** [comment]
Programador Pragmatico e Programador apaixonado

**Comment by thipacheco1** [comment]
Quero compartilhar com vocês um projeto no qual venho trabalhando intensamente: a **Formação Java Completa — Do Zero à Arquitetura**.

É uma plataforma educacional web **100% gratuita**, focada na formação prática de engenheiros de software backend. O objetivo é entregar um mapa curricular completo e sem lacunas, cobrindo desde a fundação do Java até a engenharia corporativa de produção.

🔗 **Link de Acesso:** [formacao-java.vercel.app](https://formacao-java.vercel.app)

---

### 🗺️ Matriz Curricular: 5 Fases e 21 Módulos (721 Aulas Guiadas)

A formação é dividida em **5 grandes fases evolutivas** e **21 módulos**, somando **721 aulas práticas guiadas**:

#### 🟦 Fase 1: Base Java, Ferramentas e Domínio (P0, M0 – M4 | Aulas 000 a 145)
* **P0 & M0 — Método, Ambiente e Ferramentas:** Configuração profissional do ambiente (JDK 21 LTS, IntelliJ, PowerShell, Git e Docker).
* **M1 — Fundamentos Absolutos:** Sintaxe, variáveis, controle de fluxo, arrays, métodos e lógica.
* **M2 — Java Core Profundo:** JVM, modelo de memória, registros (`records`), exceptions e leitura de documentação oficial.
* **M3 & M4 — Métodos, OO e Domínio:** Quebra de responsabilidades, coesão, refatoração, encapsulamento e projetos de console baseados em domínio.

#### 🟪 Fase 2: Java Moderno, Collections e Design (M5 – M10 | Aulas 146 a 244)
* **M5 — Collections Framework:** List, Set, Map, filas e critérios de escolha de estruturas de dados.
* **M6 — Generics e Optional:** PECS, wildcards, type erasure, Optional e repositórios tipados.
* **M7 — Programação Funcional:** Lambdas, `Functional Interfaces`, Streams API e Collectors avançados.
* **M8 — Utilitários e I/O:** Leitura/escrita de arquivos, manipulação de CSV robusto, `java.time` e UUID.
* **M9 & M10 — SOLID e Design Patterns:** Princípios SOLID na prática e Padrões GoF (Strategy, Factory, Adapter, Observer, etc.) aplicados ao backend.

#### 🟩 Fase 3: Ferramentas, Banco de Dados e Spring (M11 – M14 | Aulas 245 a 410)
* **M11 — Ferramentas Backend:** Build com Maven e Gradle, testes unitários com JUnit, qualidade, Docker e CI/CD.
* **M12 — SQL e PostgreSQL:** Modelagem relacional, consultas complexas, índices, transações e diagnósticos.
* **M13 — Persistência Java:** Da origem com JDBC puro ao ORM com JPA, Hibernate, Spring Data e tuning de performance.
* **M14 — Spring Boot e REST APIs:** Desenvolvimento de APIs RESTful profissionais, DTOs, Bean Validation, tratamento global de erros e OpenAPI/Swagger.

#### 🟧 Fase 4: Segurança, Integrações e Produção (M15 – M18 | Aulas 411 a 610)
* **M15 — Segurança de Aplicações:** Spring Security, autenticação JWT, OAuth2 e mitigações OWASP.
* **M16 — Integrações e Mensageria:** APIs externas, resiliência, mensageria com RabbitMQ/Kafka, padrões SAGA e Outbox Pattern.
* **M17 — DevOps, K8s e Cloud:** Conteinerização avançada, pipelines CI/CD, Kubernetes e infraestrutura Cloud.
* **M18 — Observabilidade e Concorrência:** Logs estruturados, métricas, distributed tracing, tuning de JVM, concorrência e `ExecutorService`.

#### 🟥 Fase 5: Arquitetura, DDD e Defesa Técnica (M19 – M20 | Aulas 611 a 720)
* **M19 — Arquitetura e DDD:** Domain-Driven Design (Aggregates, Value Objects, Domain Events), sistemas distribuídos, Event Sourcing, CQRS e decisões arquiteturais.
* **M20 — Projeto Final e Carreira:** Desenvolvimento do projeto final defensável, estruturação de portfólio no GitHub, simulação de banca técnica e preparação para entrevistas backend.

---

### 💻 Como Funcionam as Aulas na Plataforma?

Cada aula na plataforma é estruturada como um **laboratório guiado por passos**:
* **Roteiro Prático:** Objetivos e cronograma claro por etapa.
* **Ambiente Interativo Mão na Massa:** Código-fonte explicativo, simulação de execução no terminal PowerShell e inspeção visual de arquivos.
* **Análise de Evidências & Clínicas de Erros:** Onde o aluno aprende a diagnosticar falhas reais de compilação, divergências de versão do JDK e exceções comuns em desenvolvimento.
* **Acompanhamento de Progresso e Certificado:** Progresso salvo com emissão de certificado oficial ao concluir a jornada.

---

### 💬 Convite à Comunidade

A plataforma é **totalmente gratuita**, aberta para estudos e sem nenhuma barreira de entrada.

Fiquem à vontade para acessar, testar os laboratórios e enviar **feedbacks, sugestões de melhoria ou dúvidas** nos comentários!

👉 **Acesse:** [formacao-java.vercel.app](https://formacao-java.vercel.app)  
👤 **Autor:** Thiago Jorge Pacheco

Bons estudos a todos! ☕🔥
